Duplicati/Library/Backend/Telegram/TelegramBackend.cs

```diff
@@ -0,0 +1,504 @@
+using Duplicati.Library.Interface;
+using Duplicati.Library.Logging;
+using TeleSharp.TL;
+using TeleSharp.TL.Channels;
+using TeleSharp.TL.Messages;
+using TLSharp.Core;
+using TLSharp.Core.Exceptions;
+using TLSharp.Core.Network.Exceptions;
+using TLSharp.Core.Utils;
+using TLRequestDeleteMessages = TeleSharp.TL.Channels.TLRequestDeleteMessages;
+
+namespace Duplicati.Library.Backend
+{
+    public class Telegram : IStreamingBackend, IBackend
+    {
+        private const int MEBIBYTE_IN_BYTES = 1048576;
+
+        private static readonly InMemorySessionStore m_sessionStore = new InMemorySessionStore();
+        private static readonly string m_logTag = Log.LogTagFromType(typeof(Telegram));
+        private static readonly object m_lockObj = new object();
+        private static TelegramClient m_telegramClient;
+
+        private readonly int m_apiId;
+        private readonly string m_apiHash;
+        private readonly string m_authCode;
+        private readonly string m_password;
+        private readonly string m_channelName;
+        private readonly string m_phoneNumber;
+        private TLChannel m_channelCache;
+
+        public Telegram()
+        { }
+
+        public Telegram(string url, Dictionary<string, string> options)
+        {
+            if (options.TryGetValue(Strings.API_ID_KEY, out var apiId))
+            {
+                m_apiId = int.Parse(apiId);
+            }
+
+            if (options.TryGetValue(Strings.API_HASH_KEY, out var apiHash))
+            {
+                m_apiHash = apiHash.Trim();
+            }
+
+            if (options.TryGetValue(Strings.PHONE_NUMBER_KEY, out var phoneNumber))
+            {
+                m_phoneNumber = phoneNumber.Trim();
+            }
+
+            if (options.TryGetValue(Strings.AUTH_CODE_KEY, out var authCode))
+            {
+                m_authCode = authCode.Trim();
+            }
+
+            if (options.TryGetValue(Strings.AUTH_PASSWORD, out var password))
+            {
+                m_password = password.Trim();
+            }
+
+            if (options.TryGetValue(Strings.CHANNEL_NAME, out var channelName))
+            {
+                m_channelName = channelName.Trim();
+            }
+
+            if (m_apiId == 0)
+            {
+                throw new UserInformationException(Strings.NoApiIdError, nameof(Strings.NoApiIdError));
+            }
+
+            if (string.IsNullOrEmpty(m_apiHash))
+            {
+                throw new UserInformationException(Strings.NoApiHashError, nameof(Strings.NoApiHashError));
+            }
+
+            if (string.IsNullOrEmpty(m_phoneNumber))
+            {
+                throw new UserInformationException(Strings.NoPhoneNumberError, nameof(Strings.NoPhoneNumberError));
+            }
+
+            if (string.IsNullOrEmpty(m_channelName))
+            {
+                throw new UserInformationException(Strings.NoChannelNameError, nameof(Strings.NoChannelNameError));
+            }
```

Comment:
If all these fields are required, we should display them by default in the UI.  For example, see the following examples for the Tardigrade backend:

* https://github.com/duplicati/duplicati/blob/v2.0.5.110-2.0.5.110_canary_2020-08-10/Duplicati/Library/Backend/Tardigrade/TardigradeConfig.cs
* https://github.com/duplicati/duplicati/pull/4197/files#diff-09aad22515a671afd5defa7c0a14c108
* https://github.com/duplicati/duplicati/blob/v2.0.5.110-2.0.5.110_canary_2020-08-10/Duplicati/Server/webroot/ngax/templates/backends/tardigrade.html

![image](https://user-images.githubusercontent.com/17345532/94083986-5086a580-fdb9-11ea-9f46-e474b94e46fc.png)